### ISSUE TITLE
gemv failures: Native API returns: -34 (PI_ERROR_INVALID_CONTEXT)

### ISSUE THREAD

**Issue by rscohn2** [oneapi-src/distributed-ranges#125]
Fails in CI. It does not fail every time. It does not fail when running manually.

[ RUN      ] SparseMatrix.Gemv
terminate called after throwing an instance of 'sycl::_V1::runtime_error'
  what():  Native API failed. Native API returns: -34 (PI_ERROR_INVALID_CONTEXT) -34 (PI_ERROR_INVALID_CON
 
Disabled in CI with #124 

Looks similar to #121

**Comment by rscohn2** [comment]
@BenBrock @lslusarczyk 
Another sighting in shp-3-only: https://github.com/oneapi-src/distributed-ranges/actions/runs/4511947761/jobs/7944746094
```
[----------] 4 tests from DistributedVector
[ RUN      ] DistributedVector.ConstructorBasic
[       OK ] DistributedVector.ConstructorBasic (28 ms)
[ RUN      ] DistributedVector.ConstructorFill
terminate called after throwing an instance of 'sycl::_V1::runtime_error'
  what():  Native API failed. Native API returns: -34 (PI_ERROR_INVALID_CONTEXT) -34 (PI_ERROR_INVALID_CONTEXT)
<end of output>
Test time =   9.03 sec
----------------------------------------------------------
Test Failed.
"shp-3-only" end time: Mar 24 14:07 UTC
"shp-3-only" time elapsed: 00:00:09
```

Here is the test: https://github.com/oneapi-src/distributed-ranges/blob/76f46c2cbf3099735c3318033e0de49e24d6bca1/test/gtest/include/common/distributed_vector.hpp#L54

**Comment by lslusarczyk** [comment]
since #131 `sycl::queue` is sometimes given `sycl::contex` we create in `init.hpp` by `internal::global_context_ = new sycl::context(internal::devices_);` and sometimes just default context.

This suspicious to me. @BenBrock , AFAIR you said that default context should be equal to the one which we created. But this is not obvious to me. Could we stick to always use our context created by `init.hpp`, which should make things clear that context is indeed the same in all queues we use?

**Comment by BenBrock** [comment]
@lslusarczyk I switched to always using our context, and that worked fine on ORTCE, but it broke a bunch of things in the CI, which is why I reverted back.  Happy to try switching back if you guys want to.

At this point I feel like there must be a bug in the SYCL CPU runtime that the CI somehow triggers.

**Comment by rscohn2** [comment]
At least 2 fails for ConstructorFill
https://github.com/oneapi-src/distributed-ranges/actions/runs/4546343268/jobs/8016318402?pr=152

**Comment by BenBrock** [comment]
Is there any way we can get an interactive console on one of these GitHub actions instances?  Would be super useful for debugging.

**Comment by rscohn2** [comment]
> Is there any way we can get an interactive console on one of these GitHub actions instances? Would be super useful for debugging.

I don't know of a way. Maybe fill is causing a buffer overrun and smashing the stack?

**Comment by rscohn2** [comment]
I captured the error with `SYCL_PI_TRACE=-1`
USMFree is failing

I attached stdout and stderr
[error.txt](https://github.com/oneapi-src/distributed-ranges/files/11152118/error.txt)
[output.txt](https://github.com/oneapi-src/distributed-ranges/files/11152119/output.txt)

**Comment by rscohn2** [comment]
Here are the logs for a successful run:
[output.good.txt](https://github.com/oneapi-src/distributed-ranges/files/11152185/output.good.txt)
[error.good.txt](https://github.com/oneapi-src/distributed-ranges/files/11152186/error.good.txt)

**Comment by BenBrock** [comment]
```
---> piEventsWait(
	<unknown> : 1
	pi_event * : 0x5e62508[ 0x4f9f0b8 ... ]
) ---> 	pi_result : PI_SUCCESS
	[out]pi_event * : 0x5e62508[ 0x4f9f0b8 ... ]

---> piEventRelease(
	pi_event : 0x4f9f0b8
) ---> 	pi_result : PI_SUCCESS

---> piQueueRelease(
	<unknown> : 0x4540e58
) ---> 	pi_result : PI_SUCCESS

---> piextUSMFree(
	<unknown> : 0x2164f68
	<unknown> : 0x21674f0
) ---> 	pi_result : PI_SUCCESS
	[out]void * : 0x21674f0

---> piextUSMFree(
	<unknown> : 0x2164f68
	<unknown> : 0x2167d50
) ---> 	pi_result : -34
	[out]void * : 0x2167d50
```

Do we know what the two `<unknown>` arguments are?  Is this a double-free, since `0x2164f68` is passed twice?

**Comment by rscohn2** [comment]
Here: https://github.com/intel/llvm/blob/74b3d45ccbe63a6e3a8df7eb046a58328ae758b7/sycl/plugins/opencl/pi_opencl.cpp#L1517
first arg is context

**Comment by rscohn2** [comment]
Likely invalid context error is detected here: https://github.com/intel/llvm/blob/74b3d45ccbe63a6e3a8df7eb046a58328ae758b7/sycl/plugins/opencl/pi_opencl.cpp#L165

**Comment by rscohn2** [comment]
The previous free succeeds on the same context, so likely something is corrupted by the earlier free. But it is strange that fill operations trigger the problem.

**Comment by rscohn2** [comment]
I am trying to capture another fail to see if the problem always happens in the same place. There was a window of time when I could easily reproduce the problem. Now I can't. I will let it run overnight.

**Comment by rscohn2** [comment]
This time it happened on the first free

[error.2.txt](https://github.com/oneapi-src/distributed-ranges/files/11152679/error.2.txt)
[output.2.txt](https://github.com/oneapi-src/distributed-ranges/files/11152680/output.2.txt)

**Comment by lslusarczyk** [comment]
> I captured the error with `SYCL_PI_TRACE=-1`

Which test case these logs are from?
It was run on latest [3c9e748](https://github.com/lslusarczyk/distributed-ranges/commit/3c9e7488a67b2bd64d613940266f7dd3f54113b2) commit, right?

**Comment by rscohn2** [comment]
Here is the script:
```
#! /bin/bash

while :
do
    set -e
    SYCL_PI_TRACE=-1 ./shp-tests --gtest_filter=Fill/0.Range 2> error.txt 1> output.txt
done
```
> It was run on latest [3c9e748](https://github.com/lslusarczyk/distributed-ranges/commit/3c9e7488a67b2bd64d613940266f7dd3f54113b2) commit, right?

Right

I looked at a few more fails. Always first or second free. 2 devices. 10 element distributed vector, so 40 bytes total. Each device gets 20. It finishes fill and then reading memory back for the check. It is freeing the data for the distributed vector, probably in the destructor. Maybe something about the order resources are destroyed and memory is freed?

**Comment by BenBrock** [comment]
Try cherry-picking `91b1ed205706bf7c1a0aa59e23c6db89c5463e08` and see if there's any effect.

**Comment by BenBrock** [comment]
You also might try commenting out line 117 in `vector.hpp` (the `allocator_.deallocate()`) and just see if that fixes it.

**Comment by mateuszpn** [comment]
Another occurence in CI tests:
```
[----------] 4 tests from Fill/0, where TypeParam = shp::distributed_vector<int, shp::device_allocator<int, 0ul> >
[ RUN      ] Fill/0.Range
terminate called after throwing an instance of 'sycl::_V1::runtime_error'
  what():  Native API failed. Native API returns: -34 (PI_ERROR_INVALID_CONTEXT) -34 (PI_ERROR_INVALID_CONTEXT)

```
 
https://github.com/oneapi-src/distributed-ranges/actions/runs/4669235033/jobs/8267401637?pr=191

**Comment by rscohn2** [comment]
Appears to be fixed by #193
